jill/download.py

```diff
@@ -142,21 +142,22 @@ def download_package(version=None, sys=None, arch=None, *,
     msg = f"downloading Julia release for {release_str}"
     logging.info(msg)
     print(msg)
-    registry = SourceRegistry(upstream=upstream)
-    url = registry.query_download_url(version, system, architecture)
-    if not url and upstream != "Official":
-        msg = f"failed to find {release_str} in upstream {upstream}. Fallback to upstream Official."
-        logging.warning(msg)
-        print(f"{color.RED}{msg}{color.END}")
-        
-        upstream = "Official"
+
+    if not (upstream == "Official" or upstream is None):
         registry = SourceRegistry(upstream=upstream)
         url = registry.query_download_url(version, system, architecture)
         if not url:
-            msg = f"failed to find {release_str} in upstream {upstream}. Please try it later."
+            msg = f"failed to find {release_str} in upstream {upstream}. Fallback to upstream Official."
             logging.warning(msg)
             print(f"{color.RED}{msg}{color.END}")
-            return None
+
+    registry = SourceRegistry(upstream=None if upstream is None else "Official")
```

Comment:
If `upstream != "Official"` and the `query_download_url` call works, i.e. `url != False` it now queries the url twice, which seems unnecessary.

Reply:
The registry is a singleton object so it only do real-query once; all other queries are directly using the caches.

Reply:
Oh! I think you're correct here.

More importantly, it always forces the real upstream to either `"Official"` or `None` even if we pass `SJTUG`:

```bash
$ python -m jill download 1.6.0 --upstream BFSU
downloading Julia release for 1.6.0-mac-x86_64
downloading from https://julialang-s3.julialang.org/bin/mac/x64/1.6/julia-1.6.0-mac64.dmg
```

The correct logic should be:

```python
if upstream == "Official or upstream is None:
    # skip step 1 and exit once failed
else:
    # do step 1 and step 2
```

Reply:
The one I made up is in 6c51ca978020baa3d75f48376272d44226ac374c